File: mpp/lab4/common/src/main/java/service/BaseService.java
package service;

import domain.Score;

import java.rmi.RemoteException;
import java.util.ArrayList;
import java.util.List;

public abstract class BaseService implements IService {
    final List<IServiceObserver> observers = new ArrayList<>();

    void observerSetScore(Score score) {
        observers.forEach(observer -> {
            try {
                observer.onSetScore(score);
            } catch (ServiceError | RemoteException serviceError) {
                serviceError.printStackTrace();
            }
        });
    }

    public void addObserver(IServiceObserver observer) {
        observers.add(observer);
    }

    public void removeObserver(IServiceObserver observer) {
        observers.remove(observer);
    }
}

File: mpp/lab7/server/Service/BaseService.cs
using System.Collections.Generic;
using server.Domain;

namespace server.Service
{
    public class BaseService
    {
        private readonly List<IServiceObserver> _observers = new List<IServiceObserver>();

        protected void ObserverSetScore(Score score) {
            _observers.ForEach(observer => {
                observer.OnSetScore(score);
            });
        }

        public void AddObserver(IServiceObserver observer)
        {
            _observers.Add(observer);
        }

        public void RemoveObserver(IServiceObserver observer)
        {
            _observers.Remove(observer);
        }
    }
}